### PR TITLE
ADL Filesystem bug fixes

### PR DESCRIPTION
**Pull request by begoldsm**
1. Issue: 123 give better 2FA
2. Issue with re-uploading the same file with overwrite=true failing
when the file is large.
3. Update help and readme to be more accurate.
4. Move Benchmarks and Cli to a new samples folder, since these are
examples of how to integrate the package into other scripts.
5. Add support for the latest version of python (3.6)

### CODE REVIEW

**Comment by begoldsm** [comment]
@lmazuel @brettcannon could you guys take a look at this and let me know if there are any additional changes I should make or issues you see?

File: azure/datalake/store/multithread.py [hunk 1]
```diff
@@ -478,10 +484,18 @@ def put_chunk(adlfs, src, dst, offset, size, buffersize, blocksize, delimiter=No
...
     try:
         # note that it is assumed that only temp files from this run are in the segment folder created.
         # so this call is optimized to instantly delete the temp folder on concat.
+        # if somehow the target file was created between the beginning of upload
+        # and concat, we will remove it if the user specified overwrite.
+        if (adlfs.exists(outfile)):
+            if (overwrite):
+                adlfs.remove(outfile)
+            else:
+                raise FileExistsError(rpath)
```
Comment (lmazuel) [inline azure/datalake/store/multithread.py]: I don't see the rpath variable?

File: azure/datalake/store/transfer.py [hunk 2]
```diff
@@ -387,7 +387,8 @@ def _update(self, future):
                     merge_future = self._submit(
                         self._merge, self._adlfs, dst,
                         [chunk for chunk, _ in sorted(cstates.objects,
-                                                      key=lambda obj: obj[1])])
+                                                      key=lambda obj: obj[1])], 
+                        overwrite=self._parent._overwrite)
```
Comment (lmazuel) [inline azure/datalake/store/transfer.py]: Arg, I don't like using "_" fields.

File: azure/datalake/store/multithread.py (hunk 1 above)
Comment (begoldsm) [inline azure/datalake/store/multithread.py]: Good catch! Fixed!

File: azure/datalake/store/transfer.py (hunk 2 above)
Comment (brettcannon) [inline azure/datalake/store/transfer.py]: As long as those fields are all private to this module it's okay, but if whatever is stored in self._parent isn't then it's bad.

File: azure/datalake/store/transfer.py [hunk 3]
```diff
@@ -387,7 +387,8 @@ def _update(self, future):
                     merge_future = self._submit(
                         self._merge, self._adlfs, dst,
                         [chunk for chunk, _ in sorted(cstates.objects,
-                                                      key=lambda obj: obj[1])])
+                                                      key=lambda obj: obj[1])],
```
Comment (brettcannon) [inline azure/datalake/store/transfer.py]: Use https://docs.python.org/3/library/operator.html#operator.itemgetter to make this cleaner.

File: azure/datalake/store/multithread.py [hunk 4]
```diff
@@ -478,10 +484,18 @@ def put_chunk(adlfs, src, dst, offset, size, buffersize, blocksize, delimiter=No
...
 
 
-def merge_chunks(adlfs, outfile, files, shutdown_event=None):
+def merge_chunks(adlfs, outfile, files, shutdown_event=None, overwrite=False):
     try:
         # note that it is assumed that only temp files from this run are in the segment folder created.
         # so this call is optimized to instantly delete the temp folder on concat.
+        # if somehow the target file was created between the beginning of upload
+        # and concat, we will remove it if the user specified overwrite.
+        if (adlfs.exists(outfile)):
```
Comment (brettcannon) [inline azure/datalake/store/multithread.py]: No parentheses around the guard expression.

File: azure/datalake/store/transfer.py (hunk 2 above)
Comment (lmazuel) [inline azure/datalake/store/transfer.py]: Two private attributes [chained ](https://en.wikipedia.org/wiki/Law_of_Demeter) makes me nervous. But yes, it's not a proof but just a warning :)

File: azure/datalake/store/multithread.py [hunk 5]
```diff
@@ -321,6 +321,12 @@ def __init__(self, adlfs, rpath, lpath, nthreads=None, chunksize=2**28,
                  delimiter=None, overwrite=False, verbose=True):
         if not overwrite and adlfs.exists(rpath):
             raise FileExistsError(rpath)
+        
+        # forcibly remove the target file before execution
+        # if the user indicates they want to overwrite the destination.
+        if overwrite and adlfs.exists(rpath):
+            adlfs.remove(rpath);
```
Comment (brettcannon) [inline azure/datalake/store/multithread.py]: Semi-colon is unnecessary.

File: azure/datalake/store/transfer.py (hunk 2 above)
Comment (begoldsm) [inline azure/datalake/store/transfer.py]: I agree with both of you here, but I don't want to have to copy the value of the top level parameter (overwrite) everywhere. This value should be considered private in that once the parent is instantiated it should not be modified, as it is used to complete the upload (or download) operation.

**Comment by brettcannon** [comment]
LGTM

**Pull request merged**